libraries/libfc/include/fc/container/ordered_diff.hpp

```diff
@@ -106,19 +109,26 @@ class ordered_diff {
    requires std::same_as<std::decay_t<X>, diff_result>
    static Container<T> apply_diff(Container<T>&& container, X&& diff) {
       // Remove from the source based on diff.remove_indexes
-      std::ptrdiff_t offset = 0;
-      for (SizeType index : diff.remove_indexes) {
-         auto updated_index = index + offset;
-         // Safe to do static_cast as `updated_index >= 0` is verified first
-         FC_ASSERT(updated_index >= 0 && (static_cast<Container<T>::size_type>(updated_index) < container.size()), "diff.remove_indexes index ${idx} + offset ${o} not in range ${s}",
-                   ("idx", index)("o", offset)("s", container.size()));
+      for (typename Container<T>::size_type i = 0; i < diff.remove_indexes.size(); ++i) {
+         FC_ASSERT(i == 0 || diff.remove_indexes[i] > diff.remove_indexes[i-1],
+                   "diff.remove_indexes not strictly monotonically increasing: current index ${c}, previous index ${p}",
+                   ("c", diff.remove_indexes[i])("p", diff.remove_indexes[i-1]));
+
+         assert(diff.remove_indexes[i] >= i);
+         assert(std::numeric_limits<typename Container<T>::size_type>::min() + i <= diff.remove_indexes[i]); // check for underflow of diff.remove_indexes[i] - i
```

Comment:
Because of the assert at line 117, there can be no underflow.
Furthermore `std::numeric_limits<typename Container<T>::size_type>::min()` is zero if the type is unsigned, which you should require for this class.